lib/chef/provisioning/fog_driver/driver.rb

```diff
@@ -424,46 +425,63 @@ def wait_for_transport(action_handler, machine_spec, machine_options, server)
     def attach_floating_ips(action_handler, machine_spec, machine_options, server)
-      # TODO this is not particularly idempotent. OK, it is not idempotent AT ALL.  Fix.
-      if option_for(machine_options, :floating_ip_pool)
+      pool = option_for(machine_options, :floating_ip_pool)
+      floating_ip = option_for(machine_options, :floating_ip)
+      if pool
         Chef::Log.info 'Attaching IP from pool'
-        action_handler.perform_action "attach floating IP from #{option_for(machine_options, :floating_ip_pool)} pool" do
-          attach_ip_from_pool(server, option_for(machine_options, :floating_ip_pool))
+        allocated_addresses = server.addresses[pool]
+        unless allocated_addresses.nil?
+          floating_ips = allocated_addresses.select {|a_info| a_info['OS-EXT-IPS:type']=='floating'}
+          if floating_ips.length>0
+            Chef::Log.info "Server already assigned floating_ips `#{floating_ips}` from pool `#{pool}`"
+          end
+        else
+          action_handler.perform_action "Attaching floating IP from pool `#{pool}`" do
+            attach_ip_from_pool(server, pool)
+          end
         end
-      elsif option_for(machine_options, :floating_ip)
+      elsif floating_ip
         Chef::Log.info 'Attaching given IP'
-        action_handler.perform_action "attach floating IP #{option_for(machine_options, :floating_ip)}" do
-          attach_ip(server, option_for(machine_options, :allocation_id), option_for(machine_options, :floating_ip))
+        allocated_addresses = []
+        server.addresses.keys.each do |pool|
+          server.addresses[pool].each {|a_info| allocated_addresses<<a_info['addr']}
+        end
+        if allocated_addresses.include? floating_ip
+          Chef::Log.info "Address <#{floating_ip}> already allocated"
+        else
+          action_handler.perform_action "Attaching floating IP #{floating_ip}" do
+            attach_ip(server, floating_ip)
+          end
         end
       end
     end
 
     # Attach IP to machine from IP pool
     # Code taken from kitchen-openstack driver
-    #    https://github.com/test-kitchen/kitchen-openstack/blob/master/lib/kitchen/driver/openstack.rb#L196-L207
+    #    https://github.com/test-kitchen/kitchen-openstack/blob/master/lib/kitchen/driver/openstack.rb
     def attach_ip_from_pool(server, pool)
-      @ip_pool_lock ||= Mutex.new
-      @ip_pool_lock.synchronize do
+      @@ip_pool_lock.synchronize do
         Chef::Log.info "Attaching floating IP from <#{pool}> pool"
-        free_addrs = compute.addresses.collect do |i|
-          i.ip if i.fixed_ip.nil? and i.instance_id.nil? and i.pool == pool
+        free_addrs = compute.addresses.map do |i|
+          i.ip if i.fixed_ip.nil? && i.instance_id.nil? && i.pool == pool
         end.compact
         if free_addrs.empty?
-          raise ActionFailed, "No available IPs in pool <#{pool}>"
+          raise RuntimeError, "No available IPs in pool <#{pool}>"
```

Comment:
I was getting a `NameError` error because it didn't know about the ActionFailed class.  It looks like that was originally copied from knife-openstack?